Initial work on contact for explicit finite element dynamics

recuero opened the pull request:
Closes #25666.

Closes #26243.

recuero committed:
Initial commit with:

 - Initial explicit dynamics action
 - Initial explicit dynamics node-face normal constraint

(#25666).

recuero committed:
Add wave speed as a material property

recuero committed:
WIP iterative contact dynamics

recuero committed:
Update calls to system base with new API

recuero committed:
Neighbor material in node-face explicit dynamics constraint

recuero committed:
Add necessary material properties

lindsayad committed:
Mods to avoid OOB access

recuero committed:
Enable (hopefully) neighbor materials in explicit contact constraint

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1884701/) on 88fb2d0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26073/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26073/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 26072ad7c861db5410ec1d1a33332808f75bdaee`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1885005/) on 5142376 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26073/site)

This comment will be updated on new commits.

recuero committed:
Implement momentum-balance explicit dynamics contact force

recuero committed:
Fix nodal calculations

recuero committed:
Make NodeFace writeable friendly.

Pull variables into constraint.

recuero committed:
Implement balance-momentum equations for explicit contact (#25666)

Kinematic constraint works for primary surface being rigid and works
also in parallel.

Requires verification and adding documentation.

dschwen commented on framework/include/constraints/NodeFaceConstraint.h:
```diff
@@ -287,6 +301,9 @@ class NodeFaceConstraint : public Constraint,
   Real _secondary_residual;
 
+  /// An empty material property dependency set for use with \p getMatPropDependencies
+  const std::set<unsigned int> _empty_mat_prop_deps;
```
Could be static

dschwen commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -978,4 +978,7 @@ class NonlinearSystemBase : public SystemBase, public PerfGraphInterface
   std::size_t _num_scaling_groups;
+
+  /// Mutex for auxiliary variables
+  static Threads::spin_mutex writable_variable_mutex;
```
Is this used anywhere?

recuero committed:
Add heavy test

recuero committed:
Remove unused gap variable

recuero committed:
Fix documentation

recuero committed:
Move test folder and remove unnecessary mutex.

Fix website.

Only update auxiliary system in node-face constraints if we are writing
to variables.

recuero committed:
Fix critical time step postprocessor threading behavior.

When multiple threads provide values for the critical values for
the time step, pick the minimum value. Otherwise:

 - It's numerically inaccurate, only the minimum value would guarantee stability.
 - The behavior would be inconsistent with parallel processing

lindsayad commented:
@recuero do you think it makes more sense for a mechanics person to review this or do you want me to review it?

recuero commented:
I believe @dschwen will want to review it. I'll be happy to address your comments too. 

I still need to address a threading diff on a critical time step output, which I don't know if it's related to the code in this PR.

lindsayad commented:
I'm happy to let @dschwen have the reins 😄

lindsayad commented:
> still need to address a threading diff on a critical time step output, which I don't know if it's related to the code in this PR.

If you want any help looking at this let me know. Is it repeatable?

recuero commented:
It seems repeatable. I haven't looked into it yet. I have a small suspicion that we may be summing the output of threads instead of just getting the minimum. I'll ping you if I can't figure it out, thanks!

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1994305/) on 2e9b910 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8ac5e2</th>
      <th colspan=4>#26073 2e9b91</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.09%</td>
      <td>85.07%</td>
      <td>-0.02%</td>
      <td>56.14%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>98549</td>
      <td>98577</td>
      <td>+28</td>
      <td>32</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17271</td>
      <td>17298</td>
      <td>+27</td>
      <td>25</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26073/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26073/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>8ac5e2</th>
      <th colspan=4>#26073 2e9b91</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.15%</td>
      <td>89.91%</td>
      <td>-0.24%</td>
      <td>87.86%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4007</td>
      <td>4463</td>
      <td>+456</td>
      <td>456</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>438</td>
      <td>501</td>
      <td>+63</td>
      <td>63</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26073/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26073/coverage/contact)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>8ac5e2</th>
      <th colspan=4>#26073 2e9b91</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.90%</td>
      <td>84.86%</td>
      <td>-0.03%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26939</td>
      <td>26939</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4792</td>
      <td>4805</td>
      <td>+13</td>
      <td>14</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26073/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26073/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26073/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 56.14% is less than the suggested 90.0%
- `contact` new line coverage rate 87.86% is less than the suggested 90.0%
- `tensor_mechanics` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -929,14 +931,25 @@ Coupleable::writableCoupledValue(const std::string & var_name, unsigned int comp
-  if (!var->hasBlocks(br->blockIDs()))
+  const auto * nfc = dynamic_cast<const NodeFaceConstraint *>(this);
+
+  //   mooseWarning("This object '", _obj->name(), "'is not of the block restrictable type.");
+
```
```suggestion
```

recuero committed:
Update contact boundary displacements with new step.

The previous strategy modified previous displacements to produce a velocity
that would fulfil the dynamic contact constraints. That resulted in very wrong
internal force state in the new step, which _created_ energy.

This commit attempts to directly write the displacement variables in the
new step.

recuero committed:
Include option to overwrite current solution.

This option is more consistent with the internal energy of the solid.

I.e., it overwrites the position of the contacting boundaries with the
result of the iterative contact algorithm. However, it sometimes results in
some sort of element inversion or impossible mappings. This needs to be
investigated/debugged. It may be due to MOOSE's order of operations.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2015504/) on da81b23 wanted to post the following:

Warning: This PR changes repo size by 6.77 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2015647/) on e73f7e0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26073/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26073/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8f969b50c7b2ab3b565d53f4d92d5343159fa6d8`

moosebuild commented:
Job [Docker moose-dev](https://civet.inl.gov/job/2015730/) on 7a6ed09 : invalidated by @recuero

recuero committed:
Remove dead code, add coverage, cleanup

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1130,9 +1137,18 @@ NonlinearSystemBase::setConstraintSecondaryValues(NumericVector<Number> & soluti
             // reinit variables on the primary element's face at the contact point
             _fe_problem.setNeighborSubdomainID(primary_elem, 0);
             subproblem.reinitNeighborPhys(primary_elem, primary_side, points, 0);
+            if (displaced)
+              _fe_problem.reinitNeighborPhys(
+                  _mesh.elemPtr(primary_elem->id()), primary_side, points, 0);
+            else
+              subproblem.reinitNeighborPhys(primary_elem, primary_side, points, 0);
```
Wow there's so much `reinitNeighborPhys` going on here 😅 In the undisplaced case, the exact same routine is going to be called twice in a row. Seems like lines 1143 and 1144 should be deleted

recuero replied:
I was testing SAM test with this clumsy code. The test did go into the displaced condition (feproblem call) and failed.

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1130,9 +1137,18 @@ NonlinearSystemBase::setConstraintSecondaryValues(NumericVector<Number> & soluti
             // reinit variables on the primary element's face at the contact point
             _fe_problem.setNeighborSubdomainID(primary_elem, 0);
             subproblem.reinitNeighborPhys(primary_elem, primary_side, points, 0);
+            if (displaced)
+              _fe_problem.reinitNeighborPhys(
+                  _mesh.elemPtr(primary_elem->id()), primary_side, points, 0);
```
So most routines in `FEProblemBase` call to the same routine in `DisplacedProblem` but this is one of the few that doesn't. And that's because physical points which fall on a given element ID on one mesh are not guaranteed to fall on the same element ID in the other mesh. For example if a domain expands and you take points along the boundary of the displaced mesh, then these points won't fall on *any* element in the undisplaced mesh. 

This is why SAM is failing ... libMesh is unable to map the physical points into the reference coordinate space of the undisplaced mesh element because the points don't lie on the element

recuero replied:
> because the points don't lie on the element

Is the difference between the two calls a _side_ restriction?

lindsayad replied:
I'm not sure what you mean. The only way I can think of to make this work is to pass the reference points from the inversion of the physical points in the first `reinitNeighborPhys` call to a second `reinitNeighborXXX` call. We allow something like this for cohesive zone modeling which has the same problem you're running into here

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1137,8 +1138,12 @@ NonlinearSystemBase::setConstraintSecondaryValues(NumericVector<Number> & soluti
+                primary_elem->dim(), FEType(), primary_elem, points, reference_points);
             _fe_problem.reinitNeighborPhys(
-                _mesh.elemPtr(primary_elem->id()), primary_side, points, 0);
+                _mesh.elemPtr(primary_elem->id()), primary_side, reference_points, 0);
```
You won't be able to use the reference points directly in this call because `reinitNeighborPhys` expects physical points. Unless you use a different API you'll need to forward map the reference points to their physical points on the undisplaced mesh

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1130,9 +1137,14 @@ NonlinearSystemBase::setConstraintSecondaryValues(NumericVector<Number> & soluti
             // reinit variables on the primary element's face at the contact point
             _fe_problem.setNeighborSubdomainID(primary_elem, 0);
             subproblem.reinitNeighborPhys(primary_elem, primary_side, points, 0);
+            _fe_problem.reinitNeighborPhys(
+                _mesh.elemPtr(primary_elem->id()), primary_side, points, 0);
```
will need to fix

recuero replied:
Definitely!

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1280,6 +1304,9 @@ NonlinearSystemBase::constraintResiduals(NumericVector<Number> & residual, bool
             // reinit variables on the primary element's face at the contact point
             _fe_problem.setNeighborSubdomainID(primary_elem, 0);
             subproblem.reinitNeighborPhys(primary_elem, primary_side, points, 0);
+            _fe_problem.reinitNeighborPhys(
+                _mesh.elemPtr(primary_elem->id()), primary_side, points, 0);
```
will need to fix

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -2174,8 +2273,14 @@ NonlinearSystemBase::constraintJacobians(bool displaced)
             // reinit variables on the primary element's face at the contact point
             _fe_problem.setNeighborSubdomainID(primary_elem, 0);
             subproblem.reinitNeighborPhys(primary_elem, primary_side, points, 0);
+            _fe_problem.reinitNeighborPhys(
+                _mesh.elemPtr(primary_elem->id()), primary_side, points, 0);
```
will need to fix

recuero committed:
Reinit material for contact constraints on undisplaced mesh.

This is needed for explicit dynamics contact constraints which use
material properties on both secondary node and primary projection point.

recuero committed:
Clean up and add coverage for 'overwrite' option

recuero commented:
@dschwen @lindsayad This should be ready for review.

lindsayad commented:
My biggest concern is the poor line coverage in the framework. This reflects the leaking of the contact module into the framework. Is there something we can do about this? Perhaps add some virtual methods in `NonlinearSystemBase` and create a `NonlinearSystem` derived class in the contact module? That's just a spitball.

My smaller concern is the proliferation of writable variables through different systems in MOOSE

recuero commented:
> Perhaps add some virtual methods in NonlinearSystemBase and create a NonlinearSystem derived class in the contact module?

That'll fix the coverage issue. We'd need to create that contact-specific nonlinear system from the framework, I believe. Maybe with `CONTACT_ENABLED` macros? There may be a better solution, I don't like the one I just thought of.

An alternative way of looking at this is that contact, just now, requires additional, general capabilities. That 'overwrite' method is probably not that different in nature from presetting Dirichlet BCs? Maybe cooking up a test in the framework to cover that code would be enough?

lindsayad commented:
> That'll fix the coverage issue. We'd need to create that contact-specific nonlinear system from the framework, I believe. Maybe with `CONTACT_ENABLED` macros? There may be a better solution, I don't like the one I just thought of.

Definitely don't want to do macros. I would be in favor of something more `_factory` and `MooseObject` based, e.g. make Systems moose objects that can be easily derived from and buildable by the factory. This would allow users to create arbitrary Systems in their modules/applications without the framework needing the details, just like our other systems (using lower case systems here to mean systems like Kernels, BoundaryConditions, etc.)

> An alternative way of looking at this is that contact, just now, requires additional, general capabilities. That 'overwrite' method is probably not that different in nature from presetting Dirichlet BCs? Maybe cooking up a test in the framework to cover that code would be enough?

Yes if you can make framework level tests for the framework code you've added then I would be supportive of keeping the new code in the framework

recuero committed:
Create framework test to cover overwriting of solution vector.

Test object encapsulates basic functionality that allows a node
face constraint to overwrite the solution with an explicit integrator
in the framework. This feature is used in the contact module.

recuero committed:
Convert mat prop dependencies to unordered set

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2068237/) on 25644b9 : invalidated by @recuero

lindsayad committed:
NodeFaceConstraint updates

- More doxygen
- Reorder data members for public-protected-private
- friend class NonlinearSystemBase
- No more public penetration info

lindsayad committed:
Add reinitNodeFaceFace method to NonlinearSystemBase

And remove the getNeighborPoints method

lindsayad committed:
No extra close in ActuallyExplicitEuler

Also no need for overwriteContact to be virtual

lindsayad commented on framework/src/timeintegrators/ActuallyExplicitEuler.C:
```diff
@@ -95,6 +95,8 @@ ActuallyExplicitEuler::solve()
   *_nonlinear_implicit_system->solution = _nl.solutionOld();
   *_nonlinear_implicit_system->solution += _solution_update;
 
+  _nl.overwriteContact(*_nonlinear_implicit_system->solution);
```
Comment here would be appreciated about why this is needed. Also could we re-name this method? Maybe something like `overwriteNodeFace`? Of course that still begs the question of why node-face is special. Perhaps the added comment would explain why contact/node-face is special

recuero replied:
Added a comment and renamed it. The reason why we are doing something like that for contact with explicit dynamics is that we are solving a surrogate of the problem in an uncoupled way. Once that's done, we can 'preset' or 'overwrite' whatever values the displacement or velocity values had.

lindsayad committed:
reinitNodeFaceFace -> reinitNodeFace

This after moving calls to `_fe_problem.reinitNodeFace` into the
`NonlinearSystemBase` method such that `NonlinearSystemBase::reinitNodeFace`
now does all the reinit'ing necessary for node face constraints

lindsayad committed:
Turn off debug/valgrind for slow contact petrov-galerkin tests

lindsayad commented:
It seems like the balance tests might be a little long for dbg. Might be an intermittent failure:
```
[299.9s]  TIMEOUT contact/test:explicit_dynamics.test_balance FAILED (TIMEOUT) [previous results: TIMEOUT,FINISHED]
[300.0s]  TIMEOUT contact/test:explicit_dynamics.test_balance_overwrite FAILED (TIMEOUT) [previous results: TIMEOUT,FINISHED]
```

lindsayad commented:
@recuero please review my commits and see what you think. I'll see how testing does. If something fails from my changes I will fix it

I've reviewed the framework pieces. @dschwen I'll let you review the contact pieces

recuero committed:
Address Alex's comments

recuero commented:
Looks great to me, nice cleanup!

dschwen commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1084,6 +1085,65 @@ NonlinearSystemBase::enforceNodalConstraintsJacobian()
+                                    const bool displaced)
+{
+  auto & subproblem = displaced ? static_cast<SubProblem &>(*_fe_problem.getDisplacedProblem())
+                                : static_cast<SubProblem &>(_fe_problem);
```
The `displaced == false` version of subproblem is never used. @lindsayad do we have design docs on when to use the undisplaced problem even if we have a displaced problem? I think in theory we are duplicating a lot of the bookkeeping in the two problems, but I'm guessing we only care about the undisplaced problem being up-to-date? Is there anything in the code that prevents devs from accidentally accessing state that is not being maintained on the displaced problem?

lindsayad replied:
It wasn't until this PR that we started doing dedicated reinits on `_fe_problem`. Previously we just used the `subproblem` corresponding to the value of `displaced`. I presume that @recuero made this change because he wanted to use some kind of evaluations (materials maybe?) from the undisplaced problem even in displaced node-face constraints

lindsayad replied:
My concern now, which I wasn't really that motivated to voice until your comment here, is the *reverse*. If we are going to allow use of undisplaced calculations in displaced node-face constraints, why not allow the reverse? E.g. perhaps we should also always reinit the displaced problem even if `displaced == false`

lindsayad replied:
I should be more clear. We used to only do `subproblem.reinitNeighborPhys`. (We have always done _fe_problem.reinitNodeFace`). The change that this PR introduced is that we now always do `_fe_problem.reinitNeighborPhys` and then if `displaced == true` we also do `subproblem.reinitNeighborPhys` (corresponding to the displaced problem)

recuero replied:
>  I presume that @recuero made this change because he wanted to use some kind of evaluations (materials maybe?) from the undisplaced problem even in displaced node-face constraints

That's exactly right.

The case for doing displaced mesh evaluations for an undisplaced object is less straightforward to me. I.e. maybe less likely to happen? The motivation here was to avoid failed mappings.

lindsayad replied:
Yes, I agree it's probably not very likely

dschwen commented on framework/src/timeintegrators/ActuallyExplicitEuler.C:
```diff
@@ -95,6 +95,12 @@ ActuallyExplicitEuler::solve()
 
+  // Constraints may be solved in an uncoupled way. For example, momentum-balance equations may be
+  // solved node-wise and then the solution (e.g. velocities or positions)can be applied to those
+  // nodes without solve such constraints on a system level. This strategy is being used for
```
```suggestion
  // nodes without solving for such constraints on a system level. This strategy is being used for
```
?

dschwen commented on modules/contact/src/actions/ExplicitDynamicsContactAction.C:
```diff
@@ -0,0 +1,442 @@
+#include "libmesh/petsc_nonlinear_solver.h"
+#include "libmesh/string_to_enum.h"
+
+// Counter for naming auxiliary kernels
+static unsigned int ed_contact_auxkernel_counter = 0;
```
I think the preferred pattern is to build the object names to include the action name to disambiguate. This leads to more informative object names and obviates the need for statics here. But I can make that change later on.

dschwen commented on modules/contact/src/actions/ExplicitDynamicsContactAction.C:
```diff
@@ -0,0 +1,442 @@
+          params.set<std::vector<VariableName>>("mapped_primary_gap_offset") = {
+              getParam<VariableName>("mapped_primary_gap_offset")};
+        params.set<bool>("use_displaced_mesh") = true;
+        std::string name = _name + "_contact_" + Moose::stringify(ed_contact_auxkernel_counter++);
```
I see you are using the action name. Then the counter could be a local variable. Ne need for it to be a global static.

recuero committed:
Address Daniel's review

lindsayad commented:
thanks for your great contributions @recuero
